pkg/infra/benchmark_test.go

```diff
@@ -0,0 +1,77 @@
+package infra
+
+import (
+	"net"
+	"testing"
+
+	"github.com/guoger/tape/e2e/mock"
+	"github.com/hyperledger/fabric-protos-go/peer"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+)
+
+func StartMockPeer() (*mock.Server, string) {
+	lis, _ := net.Listen("tcp", "127.0.0.1:0")
+	grpcServer := grpc.NewServer()
+	mockPeer := &mock.Server{GrpcServer: grpcServer, Listener: lis}
+	go mockPeer.Start()
+	return mockPeer, lis.Addr().String()
+}
+
+func StartProposer(signed, processed chan *Elements, done chan struct{}, logger *log.Logger, threshold int, addr string) {
+	peer := Node{
+		Addr: addr,
+	}
+	Proposer, _ := CreateProposer(peer, logger)
+	go Proposer.Start(signed, processed, done, threshold)
+}
+
+func GetData() *Elements {
+	return &Elements{
+		SignedProp: &peer.SignedProposal{},
+	}
+}
+
+// suggestion mainter checkout pr and test locally to verify the preformance benchmark.
+// for any preformance impacts features.
+// go test --bench=. -v ./... --count=1
+
+// preformance history, just logging previous time running the benchmark below at local.
+// 2020-Nov | Sam | 113753 ns/op	    9226 B/op	     184 allocs/op
+func benchmarkNPeer(concurrent int, b *testing.B) {
+	processed := make(chan *Elements, 10)
+	done := make(chan struct{})
+	signeds := make([]chan *Elements, concurrent)
+	for i := 0; i < concurrent; i++ {
+		signeds[i] = make(chan *Elements, 10)
+		mockpeer, mockpeeraddr := StartMockPeer()
+		StartProposer(signeds[i], processed, done, nil, concurrent, mockpeeraddr)
+		defer mockpeer.Stop()
+	}
+
+	b.ResetTimer()
+	go func() {
+		for i := 0; i < b.N; i++ {
+			data := &Elements{SignedProp: &peer.SignedProposal{}}
+			for _, s := range signeds {
+				s <- data
+			}
+		}
+	}()
+	var n int
+	for n < b.N {
+		<-processed
+		n++
+	}
+	b.StopTimer()
+	done <- struct{}{}
+}
+
+func BenchmarkPeerEndorsement1(b *testing.B)  { benchmarkNPeer(1, b) }
+func BenchmarkPeerEndorsement2(b *testing.B)  { benchmarkNPeer(2, b) }
+func BenchmarkPeerEndorsement3(b *testing.B)  { benchmarkNPeer(3, b) }
+func BenchmarkPeerEndorsement4(b *testing.B)  { benchmarkNPeer(4, b) }
+func BenchmarkPeerEndorsement7(b *testing.B)  { benchmarkNPeer(7, b) }
+func BenchmarkPeerEndorsement8(b *testing.B)  { benchmarkNPeer(8, b) }
+func BenchmarkPeerEndorsement11(b *testing.B) { benchmarkNPeer(11, b) }
+func BenchmarkPeerEndorsement16(b *testing.B) { benchmarkNPeer(16, b) }
```

Comment:
i know this is much more handy now :P but let's not abuse. 1248 would be enough already

Reply:
```
pkg: github.com/guoger/tape/pkg/infra
BenchmarkPeerEndorsement1
BenchmarkPeerEndorsement1-16               10000            102093 ns/op
BenchmarkPeerEndorsement2
BenchmarkPeerEndorsement2-16               10000            110707 ns/op
BenchmarkPeerEndorsement3
BenchmarkPeerEndorsement3-16                7392            153278 ns/op
BenchmarkPeerEndorsement4
BenchmarkPeerEndorsement4-16                5721            209858 ns/op
BenchmarkPeerEndorsement7
BenchmarkPeerEndorsement7-16                3198            392468 ns/op
BenchmarkPeerEndorsement8
BenchmarkPeerEndorsement8-16                2797            443755 ns/op
BenchmarkPeerEndorsement11
BenchmarkPeerEndorsement11-16               2190            575793 ns/op
BenchmarkPeerEndorsement16
BenchmarkPeerEndorsement16-16               1578            736280 ns/op
```

```
goos: darwin
goarch: amd64
pkg: github.com/guoger/tape/pkg/infra
BenchmarkPeerEndorsement1
BenchmarkPeerEndorsement1-16               12670             95373 ns/op
BenchmarkPeerEndorsement2
BenchmarkPeerEndorsement2-16               10000            102050 ns/op
BenchmarkPeerEndorsement3
BenchmarkPeerEndorsement3-16                7898            139679 ns/op
BenchmarkPeerEndorsement4
BenchmarkPeerEndorsement4-16                5822            190260 ns/op
BenchmarkPeerEndorsement7
BenchmarkPeerEndorsement7-16                3450            362334 ns/op
BenchmarkPeerEndorsement8
BenchmarkPeerEndorsement8-16                2907            420049 ns/op
BenchmarkPeerEndorsement11
BenchmarkPeerEndorsement11-16               2318            521113 ns/op
BenchmarkPeerEndorsement16
BenchmarkPeerEndorsement16-16               1636            695947 ns/op
PASS
```
I am agree with 1248 is enough already, but I am curious about the reason for my data above. (for round of 2)

Reply:
right and we could discuss it. but for this PR to get merged, let's do 1248

Reply:
```
pkg: github.com/guoger/tape/pkg/infra
BenchmarkPeerEndorsement1-16     	   10000	    109258 ns/op
BenchmarkPeerEndorsement2-16     	   10000	    122061 ns/op
BenchmarkPeerEndorsement4-16     	    5882	    230032 ns/op
BenchmarkPeerEndorsement8-16     	    2972	    449138 ns/op
BenchmarkPeerEndorsement16-16    	    1502	    737675 ns/op
BenchmarkPeerEndorsement32-16    	     978	   1369804 ns/op
BenchmarkPeerEndorsement64-16    	     382	   3213606 ns/op
```

Reply:
updated to 1248